src/System.Diagnostics.TraceSource/tests/CorrelationManagerTests.cs

```diff
@@ -0,0 +1,65 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Xunit;
+using System.Diagnostics.Tracing;
+using System.Collections;
+
+namespace System.Diagnostics.TraceSourceTests
+{
+    public class CorrelationManagerTests
+    {
+        [Fact]
+        public void CorrelationManager_CheckStack()
+        {
+            Trace.CorrelationManager.StartLogicalOperation("one");
+            Trace.CorrelationManager.StartLogicalOperation("two");
+
+            // 2 operations in progress
+            ValidateStack(Trace.CorrelationManager.LogicalOperationStack, "two", "one");
+
+            Trace.CorrelationManager.StopLogicalOperation();
+            // 1 operation in progress
+            ValidateStack(Trace.CorrelationManager.LogicalOperationStack, "one");
+
+            Trace.CorrelationManager.StopLogicalOperation();
+            // 0 operation in progress
+            ValidateStack(Trace.CorrelationManager.LogicalOperationStack, Array.Empty<object>());
+        }
+
+        [Fact]
+        public void CorrelationManager_NullOperationId()
+        {
+            Assert.Throws<ArgumentNullException>("operationId", () => Trace.CorrelationManager.StartLogicalOperation(null));
+        }
+
+        [Fact]
+        public void CorrelationManager_EmptyStack()
+        {
+            Assert.Throws<InvalidOperationException>(() => Trace.CorrelationManager.StopLogicalOperation());
+        }
+
+        [Fact]
+        public void CorrelationManager_ActivityId()
+        {
+            Trace.CorrelationManager.ActivityId = Guid.Empty;
+
+            Assert.Equal(Guid.Empty, Trace.CorrelationManager.ActivityId);
+        }
+
+        private static void ValidateStack(Stack input, params object[] expectedContents)
+        {
+            Assert.Equal(expectedContents.Length, input.Count);
+            // Note: this reverts the stack 
+            Stack currentStack = new Stack(input);
+
+            // The expected values are passed in in the order they are supposed to be in the original stack
+            // so we need to match them from the end of the array since the stack is also reversed
+            for (int i = expectedContents.Length - 1; i >= 0; i--)
+            {
+                Assert.Equal(expectedContents[i], currentStack.Pop());
+            }
+        }
+    }
```

Comment:
It'd be good to have a test that validates the information flows across async operations, e.g.
```C#
Guid g = Guid.NewGuid();
Trace.CorrelationManager.ActivityId = g;
Assert.Equal(g, Trace.CorrelationManager.ActivityId);
await Task.Yield();
Assert.Equal(g, Trace.CorrelationManager.ActivityId);
await Task.Delay(1);
Assert.Equal(g, Trace.CorrelationManager.ActivityId);
Parallel.For(0, 1000, i =>
{
    Assert.Equal(g, Trace.CorrelationManager.ActivityId);
});
```
Etc.